src/workerd/jsg/async-context.c++

```diff
@@ -187,17 +188,14 @@ void AsyncContextFrame::jsgVisitForGc(GcVisitor& visitor) {
   }
 }
 
-void IsolateBase::pushAsyncFrame(AsyncContextFrame& next) {
-  asyncFrameStack.add(&next);
-}
-
-void IsolateBase::pushRootAsyncFrame() {
-  asyncFrameStack.add(RootAsyncContextFrame{});
-}
-
-void IsolateBase::popAsyncFrame() {
-  KJ_DASSERT(asyncFrameStack.size() > 0, "the async context frame stack was corrupted");
-  asyncFrameStack.removeLast();
+void IsolateBase::setCurrentAsyncContextFrame(kj::Maybe<AsyncContextFrame&> maybeFrame) {
+  KJ_IF_MAYBE(frame, maybeFrame) {
+    maybeCurrentAsyncContextFrame = *frame;
```

Comment:
Won't `maybeCurrentAsyncContextFrame` become out-of-sync with the continuation context as tracked by V8 whenever a continuation runs? I think you always have to call V8's `GetContinuationPreservedEmbedderData()`, you can't track this separately.

It looks like the one place that reads `maybeCurrentAsyncContextFrame` is in `Scope`'s constructor, to initialize `prior`. I think at present what will happen is the outermost `Scope` created inside a promise continuation will incorrectly capture `prior` as `null`, which means than when exiting the scope, it'll discard the context instead of returning it to the correct previous value. Can we add a test for this bug?

Reply:
> Won't maybeCurrentAsyncContextFrame become out-of-sync with the continuation context as tracked by V8 whenever a continuation runs? I think you always have to call V8's GetContinuationPreservedEmbedderData(), you can't track this separately.

I've not seen it go out of sync in any of the testing I've done but you're definitely right that just calling GetContinuationPreservedEmbedderData is simpler than tracking this here and is one less thing that could go wrong just in case.